Port WebCompiledContentRuleListData to the new IPC serialization format

cdumez opened the pull request:
#### 5783584ba3aa701f6248ca97a68f5c4788af7288
<pre>
Port WebCompiledContentRuleListData to the new IPC serialization format
<a href="https://bugs.webkit.org/show_bug.cgi?id=264538">https://bugs.webkit.org/show_bug.cgi?id=264538</a>

Reviewed by Alex Christensen.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/NetworkProcess/NetworkContentRuleListManager.cpp:
(WebKit::NetworkContentRuleListManager::addContentRuleLists):
* Source/WebKit/Shared/WebCompiledContentRuleList.cpp:
(WebKit::WebCompiledContentRuleList::create):
* Source/WebKit/Shared/WebCompiledContentRuleList.h:
* Source/WebKit/Shared/WebCompiledContentRuleListData.cpp:
(WebKit::WebCompiledContentRuleListData::createDataHandle const):
(WebKit::WebCompiledContentRuleListData::WebCompiledContentRuleListData):
(WebKit::WebCompiledContentRuleListData::encode const): Deleted.
(WebKit::WebCompiledContentRuleListData::decode): Deleted.
* Source/WebKit/Shared/WebCompiledContentRuleListData.h:
* Source/WebKit/Shared/WebCompiledContentRuleListData.serialization.in: Added.
* Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
(API::createExtension):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/UserContent/WebUserContentController.cpp:
(WebKit::WebUserContentController::addContentRuleLists):

Canonical link: <a href="https://commits.webkit.org/270505@main">https://commits.webkit.org/270505@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a3ab3a1d1ac7cdc3a322c3f058f352a302a284a0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25677 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4282 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26960 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27777 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23517 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6031 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1717 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23643 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25926 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3194 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22127 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28357 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/2822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/29157 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23430 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23447 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/27015 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2840 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1076 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4222 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3290 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3282 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3156 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Shared/WebCompiledContentRuleListData.cpp:
```diff
@@ -32,97 +32,31 @@
-        WTFMove(*frameURLFiltersBytecodeOffset),
-        WTFMove(*frameURLFiltersBytecodeSize)
-    }};
+    if (data && data->size() < ruleListDataSize(topURLFiltersBytecodeOffset, topURLFiltersBytecodeSize))
```
An ASSERT_NOT_REACHED() may be appropriate here.
We should probably also add checks that all the offset+size combinations are in range of the data.

achristensen07 commented on Source/WebKit/Shared/WebCompiledContentRuleListData.cpp:
```diff
@@ -32,97 +32,31 @@
-    if (!*handle)
-        return std::nullopt;
-    auto data = SharedMemory::map(WTFMove(**handle), SharedMemory::Protection::ReadOnly);
-    if (!data)
```
Crash logs actually indicate this was a cause of extremely rare "decode failure" crashes when the mapping fails.  This will probably fix those crashes too.

cdumez committed:
Port WebCompiledContentRuleListData to the new IPC serialization format
https://bugs.webkit.org/show_bug.cgi?id=264538

Reviewed by Alex Christensen.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/NetworkProcess/NetworkContentRuleListManager.cpp:
(WebKit::NetworkContentRuleListManager::addContentRuleLists):
* Source/WebKit/Shared/WebCompiledContentRuleList.cpp:
(WebKit::WebCompiledContentRuleList::create):
* Source/WebKit/Shared/WebCompiledContentRuleList.h:
* Source/WebKit/Shared/WebCompiledContentRuleListData.cpp:
(WebKit::WebCompiledContentRuleListData::createDataHandle const):
(WebKit::WebCompiledContentRuleListData::WebCompiledContentRuleListData):
(WebKit::WebCompiledContentRuleListData::encode const): Deleted.
(WebKit::WebCompiledContentRuleListData::decode): Deleted.
* Source/WebKit/Shared/WebCompiledContentRuleListData.h:
* Source/WebKit/Shared/WebCompiledContentRuleListData.serialization.in: Added.
* Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
(API::createExtension):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/UserContent/WebUserContentController.cpp:
(WebKit::WebUserContentController::addContentRuleLists):

Canonical link: https://commits.webkit.org/270505@main
